feat: Update avatar props and background color calculation

kittimiyo opened the pull request:
Makes several updates to the Avatar component:
 - Adds new optional props: firstName and lastName. This gives devs the option to pass in the first and last names of the user separately, which helps avoid errors due to parsing a full name. For example, for the name "Yu Jye Foo", the name can be passed in properly separated, so that the Avatar does not need to guess about where the first name ends and the last name begins.
 - Generates initials from name, regardless of the characters in the name. The current Avatar tends to skip non-Roman alphabet characters when building initials. This means that users with non-English names will not see the correct initials, sometimes no initials are shown.
 - Background color is generated regardless of characters in the name. The current Avatar calculates the background color assuming that only Roman alphabet characters will be used, which means that users with non-English names will all have the same default background color. This change uses modulo to generate a number between 1 and 7 (inclusive), no matter what characters are in use.
 - Background color is generated based on full name. The current Avatar calculates the background color using only the first letter in the first name as a reference. This means that users with the same initials will always have the same background color. This update uses the user's full name to calculate background color, meaning that the chance of users with the same initials having the same background color is only 1 in 7.
 - Adds new optional props: label and imageAlt - there will overwrite the Avatar label and the image's alt text if provided. This way localized strings can be used.

Some examples of the updated Avatar component in storybook:
![Screen Shot 2020-11-17 at 2 56 19 PM](https://user-images.githubusercontent.com/3441699/99470786-2e615d80-28fa-11eb-9a79-e59b4b69344d.png)
![Screen Shot 2020-11-17 at 2 52 47 PM](https://user-images.githubusercontent.com/3441699/99470792-328d7b00-28fa-11eb-9318-68221adfdf0a.png)
![Screen Shot 2020-11-17 at 2 52 11 PM](https://user-images.githubusercontent.com/3441699/99470808-391bf280-28fa-11eb-984a-02ec01329471.png)
![Screen Shot 2020-11-17 at 2 59 10 PM](https://user-images.githubusercontent.com/3441699/99470834-45a04b00-28fa-11eb-80f8-fab8ce9d6e8c.png)

wmui51 commented:
@kittimiyo thanks for the pr, I will try and pull it down to test it out as soon as I can, it might not be for another week or two before I can really take a look, but I do have some quick comments.

Your git commit is a `refactor` but this I think this is more of a `feat` commit as you are adding new features and new `props`. Also as a `refactor` commit it will not trigger a release from `semantic-release`.

This is also a public repo on github so I don't think it is a good idea to reference internal Autodesk jira tickets.

There appear to be some snapshot tests that are failing from the `react-test`, you don't have to worry about triggering the `visual-test` yet as I can run that for you later.

> Generates initials from name, regardless of the characters in the name. The current Avatar tends to skip non-Roman alphabet characters when building initials. This means that users with non-English names will not see the correct initials, sometimes no initials are shown. (See https://jira.autodesk.com/browse/ACWB-8049)

For this, I think you should have a chat with @Guobi as he fixed that with this PR: https://github.com/Autodesk/hig/pull/2302 and also you may want to discuss with him on what it should be doing when localized as he mentions something about only returning 1 Asian full-width character, and from your commit it looks like you removed the Regex for this: https://github.com/Autodesk/hig/blob/development/packages/avatar/src/Avatar.js#L42.

kittimiyo commented:
Thank you Wilbert, I'll update my pull request to be a `feat`, and remove the references to Jira tickets. 
For @Guobi's changes, it looks like only part of his code was merged: https://github.com/Autodesk/hig/commit/a844ede90f21a288f9829dc1f03a67d94211ec35#diff-ca60c1e29eeae354e1f4419465612133cb5b1acd51e7a2a944402575caa99d8e
I'll talk with him so we can collaborate on the best way to get initials.

wmui51 commented:
@Guobi's PR was merged in: https://github.com/Autodesk/hig/pull/2307 and that one `regex` line was the main change.

kittimiyo commented:
Hi Wilbert, regarding the `react-test` snapshot failures, I’ve reviewed the changes and I believe they are all expected changes, since I updated the way the background color is calculated. I ran `yarn run test-ci -u` locally as the CircleCi output suggested, but it didn’t seem to make any changes to the code, so I’m not sure how to fix that error.

pardoman commented:
Would be great to have a unit test that validates the initials pulled from names in different languages. That way we can ensure that if the implementation keeps evolving, then the results will continue to remain consistent.

Test to expand: https://github.com/Autodesk/hig/blob/development/packages/avatar/src/Avatar.test.js#L46-L50

kittimiyo commented:
@pardoman great point! I'll add some tests

kittimiyo commented:
@pardoman Added some unit tests for my changes.
@wmui51 I talked with @Guobi and he is fine with my changes, he asked that I add some tests for non-English names: `Λεωνίδας ωή, Иван Виктор, 安倍晋三` I added them to my tests.

pardoman commented on packages/avatar/src/Avatar.test.js:
```diff
@@ -42,11 +42,186 @@ describe("avatar/Avatar", () => {
     }
   ]);
 
+  describe("Creating initials from name", () => {
```
Thanks. These are great!

wmui51 commented on packages/avatar/src/Avatar.js:
```diff
@@ -165,14 +222,31 @@ class Avatar extends Component {
 
   render() {
     const { size, name, ...otherProps } = this.props;
+    const { firstName, lastName } = otherProps;
```
I think these should come from `this.props` above rather than `otherProps` since these are actual declared props for the component. Also `className` is already being destructured from `otherProps`.

kittimiyo replied:
Moved `firstName` and `lastName` declarations to `this.props` destructuring

wmui51 commented on packages/avatar/src/Avatar.js:
```diff
@@ -165,14 +222,31 @@ class Avatar extends Component {
-        ? backgroundIdFromName(name)
-        : COLOR_VARIANT_COUNT;
-    const label = `Avatar for ${name}`;
+    let nameObj = {};
```
Rather than declaring an empty object with `let` and then the `if/else` statement below, I think you could instead use `const` and then a ternary to define it based on your `(!firstName && !lastName)` condition.

kittimiyo replied:
Good point, updated.

wmui51 commented:
@kittimiyo For the failed `react-test` it looks like it is failing in the `@hig/profile-flyout` and `@hig/top-nav` components as those have `@hig/avatar` as a dependency. So you will have to update the the snapshots in those packages. When you do that keep in mind each package should get its own commit as we don't want to trigger a release by having those commits lumped together with a `feat` commit in `@hig/avatar`.

In addition, for your commits you can leave it as `feat` alone and no need to include `(Avatar)` since it is in the `@hig/avatar` package. Please squash your commits as they are both `feat` commits and each commit message will appear in the CHANGELOG and the second one seems to be a refactor of your original code and adding of tests which probably don't need to appear in the CHANGELOG. If you do want to separate your commits, I would consider lumping all the feature code together and then separate the test code with a `test` commit.

kittimiyo commented:
@wmui51
 - I squashed my changes to the Avatar and its tests into 1 commit, and removed the (Avatar) scope from the commit.
 - I updated the failing snapshots in /top-nav and /profile-flyout and added each of the changes as their own `test` commits - one for /top-nav and one for profile-flyout. Please let me know if this is the correct format.

wmui51 commented:
@kittimiyo commits and everything look good, I'll pull down and test and run the visual-test next week. I get the feeling the visual-test may have to be updated since the background color calculation has changed. Ping me if you want to know how to run the update for those, or I can run it if need be.

wmui51 commented:
@kittimiyo I just took a look at this in `storybook` and everything seems to be working as expected. The one thing I noticed was that I think we should leave `name` as the primary `prop` for the initials and leave `firstName` and `lastName` as optional in case folks need more granularity or have more edge-cases where they need the separation. When testing, I tried using a name with Chinese characters and when I separated it in `firstName` and `lastName` then we get two characters in the Avatar which is opposite of this recommendation:

https://github.com/Autodesk/hig/pull/2302

At the very least we should at least add back `name` to the `storybook` and remove "legacy" from the description in the props table.

wmui51 commented:
This looks good, pulled all commits into another branch and created a PR there (https://github.com/Autodesk/hig/pull/2348) to trigger the `visual-test`.
